File: problems/6-zigzag-conversion/python/6_zigzag_conversion.py
from typing import List


class Solution:
    def convert(self, s: str, rows: int) -> str:
        if rows == 1:
            return s

        n: int = len(s)
        matrix: List[List[chr]] = [[] for _ in range(min(rows, n))]

        zigzag: bool = False
        row: int = 0

        for c in s:
            matrix[row].append(c)

            if row == 0 or row == rows - 1:
                zigzag = not zigzag

            row += 1 if zigzag else -1

        result: List[chr] = []

        for r in matrix:
            result.extend(r)

        return "".join(result)


def test(s: str, rows: int, expected) -> None:
    result: str = Solution().convert(s, rows)
    if result == expected:
        print(f" + Test 'Convert({s}, {rows}) == {expected}' passed!")
    else:
        print(
            f" - Test 'Convert({s}, {rows}) == {expected}' failed with value {result}!"
        )


if __name__ == "__main__":
    test("PAYPALISHIRING", 3, "PAHNAPLSIIGYIR")
    test("PAYPALISHIRING", 4, "PINALSIGYAHRPI")
    test("A", 1, "A")
    test("ABC", 2, "ACB")
    test("AB", 2, "AB")

File: problems/6-zigzag-conversion/csharp/6_zigzag_conversion.cs
using System.Collections.Generic;
using System;

public static class Program
{
  public static void ConvertTest(string s, int numRows, string expected)
  {
    string result = new Solution().Convert(s, numRows).Trim();
    if (result.ToLower() == expected.ToLower()) { Console.WriteLine(" + Test 'Convert({0}, {1}) == {2}' passed!", s, numRows, expected); }
    else { Console.WriteLine(" - Test 'Convert({0}, {1}) == {2}' failed with value {3}!", s, numRows, expected, result); }
  }
  public static void Main(string[] args)
  {
    ConvertTest("PAYPALISHIRING", 3, "PAHNAPLSIIGYIR");
    ConvertTest("PAYPALISHIRING", 4, "PINALSIGYAHRPI");
    ConvertTest("A", 1, "A");
    ConvertTest("ABC", 2, "ACB");
    ConvertTest("AB", 1, "AB");
  }
}

public class Solution
{
  public string Convert(string s, int rows)
  {
    if (rows == 1)
      return s;

    int n = s.Length;
    List<List<char>> matrix = new List<List<char>>();

    for (int i = 0; i < Math.Min(rows, n); i++)
      matrix.Add(new List<char>());

    bool zigzag = false;
    int curRow = 0;

    foreach (char c in s)
    {
      matrix[curRow].Add(c);

      if (curRow == 0 || curRow == rows - 1)
        zigzag = !zigzag;

      curRow += zigzag ? 1 : -1;
    }

    List<char> result = new List<char>();

    foreach (List<char> row in matrix)
      result.AddRange(row);

    return string.Join("", result);
  }
}
